dockerd libnetwork panic during shutdown

Saw this in a powerpc `DockerSwarmSuite.TestSwarmClusterRotateUnlockKey` test failure:  https://jenkins.dockerproject.org/job/Docker-PRs-powerpc/2421/console

Downloading the logs, I see on one of the daemons:

```
time="2017-05-02T18:09:29.155646000Z" level=info msg="shutting down certificate renewal routine" module="node/tls" node.id=hn06lqptpi9ti0y9lfovfr1b0 node.role=swarm-manager 
time="2017-05-02T18:09:29.155672000Z" level=info msg="Stopping manager" module=node node.id=hn06lqptpi9ti0y9lfovfr1b0 
time="2017-05-02T18:09:29.155903000Z" level=debug msg="(*Agent).run exited" module="node/agent" node.id=hn06lqptpi9ti0y9lfovfr1b0 
time="2017-05-02T18:09:29.156142000Z" level=debug msg="stop transport" 
time="2017-05-02T18:09:29.156607000Z" level=info msg="Manager shut down" module=node node.id=hn06lqptpi9ti0y9lfovfr1b0 
time="2017-05-02T18:09:29.156735000Z" level=info msg="Initializing Libnetwork Agent Listen-Addr=0.0.0.0 Local-addr=127.0.0.1 Adv-addr=127.0.0.1 Data-addr= Remote-addr-list=[127.0.0.1 127.0.0.1]" 
time="2017-05-02T18:09:29.156724000Z" level=error msg="cluster exited with error: manager stopped: context canceled" 
time="2017-05-02T18:09:29.156800000Z" level=info msg="Gossip cluster hostname 04a5c2ed704d-6c4078152bb1" 
time="2017-05-02T18:09:29.156795000Z" level=debug msg="start clean shutdown of all containers with a 15 seconds timeout..." 
time="2017-05-02T18:09:29.156837000Z" level=debug msg="Encryption key 1: dc120" 
time="2017-05-02T18:09:29.156850000Z" level=debug msg="Encryption key 2: 57d28" 
time="2017-05-02T18:09:29.156855000Z" level=debug msg="start clean shutdown of cluster resources..." 
time="2017-05-02T18:09:29.156865000Z" level=debug msg="Encryption key 3: 2b579" 
panic: runtime error: index out of range

goroutine 175 [running]:
panic(0x11489bc0, 0xc420016090)
	/usr/local/go/src/runtime/panic.go:500 +0x3a8
github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).getKeys(0xc42000ae00, 0x116d9dca, 0x10, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/agent.go:269 +0x3b4
github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).agentInit(0xc42000ae00, 0xc42076c450, 0x7, 0xc420a5a8f0, 0x9, 0xc420a5a8f0, 0x9, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/agent.go:333 +0x6a8
github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).agentSetup(0xc42000ae00, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/agent.go:223 +0x59c
github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).clusterAgentInit(0xc42000ae00)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/controller.go:326 +0x184
created by github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).SetClusterProvider
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/controller.go:252 +0xd4

goroutine 1 [select]:
main.shutdownDaemon(0xc42039a600)
	/go/src/github.com/docker/docker/cmd/dockerd/daemon.go:398 +0x30c
main.(*DaemonCli).start(0xc420489e30, 0xc420489e00, 0x116e7f39, 0x17, 0xc420065800, 0xc420011090, 0xc42038c750, 0x0, 0x0)
	/go/src/github.com/docker/docker/cmd/dockerd/daemon.go:336 +0x1f88
main.runDaemon(0x11473e00, 0x116e7f39, 0x17, 0xc420065800, 0xc420011090, 0xc42038c750, 0x0, 0x0)
	/go/src/github.com/docker/docker/cmd/dockerd/docker.go:91 +0x1c0
main.newDaemonCommand.func1(0xc4200a8900, 0xc42039e460, 0x0, 0xe, 0x0, 0x0)
	/go/src/github.com/docker/docker/cmd/dockerd/docker.go:42 +0x98
github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).execute(0xc4200a8900, 0xc42000c100, 0xe, 0xe, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:646 +0x558
github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc4200a8900, 0xc4200a8900, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:742 +0x40c
github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).Execute(0xc4200a8900, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:695 +0x34
main.main()
	/go/src/github.com/docker/docker/cmd/dockerd/docker.go:118 +0x170

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_ppc64x.s:1095 +0x4

goroutine 6 [syscall]:
os/signal.signal_recv(0x1207cc60)
	/usr/local/go/src/runtime/sigqueue.go:116 +0x270
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x24
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x44

goroutine 11 [chan receive]:
github.com/docker/docker/libcontainerd.(*remote).handleConnectionChange(0xc420476340)
	/go/src/github.com/docker/docker/libcontainerd/remote_unix.go:146 +0x128
created by github.com/docker/docker/libcontainerd.New
	/go/src/github.com/docker/docker/libcontainerd/remote_unix.go:120 +0x8bc

goroutine 10 [select]:
github.com/docker/docker/vendor/google.golang.org/grpc.(*addrConn).transportMonitor(0xc4203a4b00)
	/go/src/github.com/docker/docker/vendor/google.golang.org/grpc/clientconn.go:751 +0x858
github.com/docker/docker/vendor/google.golang.org/grpc.(*ClientConn).resetAddrConn.func1(0xc4203a4b00)
	/go/src/github.com/docker/docker/vendor/google.golang.org/grpc/clientconn.go:506 +0x234
created by github.com/docker/docker/vendor/google.golang.org/grpc.(*ClientConn).resetAddrConn
	/go/src/github.com/docker/docker/vendor/google.golang.org/grpc/clientconn.go:507 +0x7b4

goroutine 36 [IO wait]:
net.runtime_pollWait(0x3fffa004d0a8, 0x72, 0x0)
	/usr/local/go/src/runtime/netpoll.go:160 +0x8c
net.(*pollDesc).wait(0xc42045c0d0, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x40
net.(*pollDesc).waitRead(0xc42045c0d0, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x3c
net.(*netFD).Read(0xc42045c070, 0xc420138000, 0x8000, 0x8000, 0x0, 0x120770a0, 0xc420016160)
	/usr/local/go/src/net/fd_unix.go:243 +0x248
net.(*conn).Read(0xc42042c000, 0xc420138000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:173 +0xd4
bufio.(*Reader).fill(0xc4202242a0)
	/usr/local/go/src/bufio/bufio.go:97 +0x1f8
bufio.(*Reader).Read(0xc4202242a0, 0xc420388038, 0x9, 0x9, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:209 +0x24c
io.ReadAtLeast(0x1206c7a0, 0xc4202242a0, 0xc420388038, 0x9, 0x9, 0x9, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:307 +0xec
io.ReadFull(0x1206c7a0, 0xc4202242a0, 0xc420388038, 0x9, 0x9, 0xc42003dbd0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:325 +0x60
github.com/docker/docker/vendor/golang.org/x/net/http2.readFrameHeader(0xc420388038, 0x9, 0x9, 0x1206c7a0, 0xc4202242a0, 0x0, 0xc400000000, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/golang.org/x/net/http2/frame.go:237 +0x88
github.com/docker/docker/vendor/golang.org/x/net/http2.(*Framer).ReadFrame(0xc420388000, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/golang.org/x/net/http2/frame.go:469 +0xd8
github.com/docker/docker/vendor/google.golang.org/grpc/transport.(*framer).readFrame(0xc420418090, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/google.golang.org/grpc/transport/http_util.go:508 +0x40
github.com/docker/docker/vendor/google.golang.org/grpc/transport.(*http2Client).reader(0xc42038e000)
	/go/src/github.com/docker/docker/vendor/google.golang.org/grpc/transport/http2_client.go:937 +0x108
created by github.com/docker/docker/vendor/google.golang.org/grpc/transport.newHTTP2Client
	/go/src/github.com/docker/docker/vendor/google.golang.org/grpc/transport/http2_client.go:205 +0xa64

goroutine 37 [select]:
github.com/docker/docker/vendor/google.golang.org/grpc/transport.(*http2Client).controller(0xc42038e000)
	/go/src/github.com/docker/docker/vendor/google.golang.org/grpc/transport/http2_client.go:1015 +0x660
created by github.com/docker/docker/vendor/google.golang.org/grpc/transport.newHTTP2Client
	/go/src/github.com/docker/docker/vendor/google.golang.org/grpc/transport/http2_client.go:235 +0x10c4

goroutine 38 [select]:
github.com/docker/docker/vendor/google.golang.org/grpc.newClientStream.func2(0x1208f4e0, 0xc42038e000, 0xc42041c000, 0xc42038e240)
	/go/src/github.com/docker/docker/vendor/google.golang.org/grpc/stream.go:204 +0x404
created by github.com/docker/docker/vendor/google.golang.org/grpc.newClientStream
	/go/src/github.com/docker/docker/vendor/google.golang.org/grpc/stream.go:224 +0xe74

goroutine 39 [select]:
github.com/docker/docker/vendor/google.golang.org/grpc/transport.(*recvBufferReader).Read(0xc42012c240, 0xc4203041f0, 0x5, 0x5, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/google.golang.org/grpc/transport/transport.go:140 +0x750
github.com/docker/docker/vendor/google.golang.org/grpc/transport.(*Stream).Read(0xc42041c000, 0xc4203041f0, 0x5, 0x5, 0xc42020ebd0, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/google.golang.org/grpc/transport/transport.go:325 +0x74
io.ReadAtLeast(0x120704a0, 0xc42041c000, 0xc4203041f0, 0x5, 0x5, 0x5, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:307 +0xec
io.ReadFull(0x120704a0, 0xc42041c000, 0xc4203041f0, 0x5, 0x5, 0xc42020e9a0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:325 +0x60
github.com/docker/docker/vendor/google.golang.org/grpc.(*parser).recvMsg(0xc4203041e0, 0x7fffffff, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/google.golang.org/grpc/rpc_util.go:231 +0xac
github.com/docker/docker/vendor/google.golang.org/grpc.recv(0xc4203041e0, 0x12087d60, 0x120e8900, 0xc42041c000, 0x0, 0x0, 0x115b4360, 0xc42005a0c0, 0x7fffffff, 0x0, ...)
	/go/src/github.com/docker/docker/vendor/google.golang.org/grpc/rpc_util.go:315 +0x40
github.com/docker/docker/vendor/google.golang.org/grpc.(*clientStream).RecvMsg(0xc42038e240, 0x115b4360, 0xc42005a0c0, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/google.golang.org/grpc/stream.go:312 +0xa8
github.com/docker/docker/vendor/github.com/docker/containerd/api/grpc/types.(*aPIEventsClient).Recv(0xc4202fa0f0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/github.com/docker/containerd/api/grpc/types/api.pb.go:2093 +0x88
github.com/docker/docker/libcontainerd.(*remote).handleEventStream(0xc420476340, 0x1208eb20, 0xc4202fa0f0)
	/go/src/github.com/docker/docker/libcontainerd/remote_unix.go:293 +0x4c
created by github.com/docker/docker/libcontainerd.(*remote).startEventsMonitor
	/go/src/github.com/docker/docker/libcontainerd/remote_unix.go:287 +0x2d4

goroutine 40 [select, locked to thread]:
runtime.gopark(0x1183a9c8, 0x0, 0x116c81f7, 0x6, 0x18, 0x2)
	/usr/local/go/src/runtime/proc.go:259 +0x19c
runtime.selectgoImpl(0xc42003ff10, 0x0, 0x18)
	/usr/local/go/src/runtime/select.go:423 +0xc74
runtime.selectgo(0xc42003ff10)
	/usr/local/go/src/runtime/select.go:238 +0x14
runtime.ensureSigM.func1()
	/usr/local/go/src/runtime/signal1_unix.go:304 +0x530
runtime.goexit()
	/usr/local/go/src/runtime/asm_ppc64x.s:1095 +0x4

goroutine 51 [chan receive]:
github.com/docker/docker/pkg/signal.Trap.func1(0xc420224540, 0xc420304300)
	/go/src/github.com/docker/docker/pkg/signal/trap.go:37 +0x74
created by github.com/docker/docker/pkg/signal.Trap
	/go/src/github.com/docker/docker/pkg/signal/trap.go:67 +0x1f4

goroutine 53 [chan receive]:
github.com/docker/docker/daemon.(*Daemon).setupDumpStackTrap.func1(0xc420386780, 0x3fffce8cfa91, 0x22, 0xc42039a600)
	/go/src/github.com/docker/docker/daemon/debugtrap_unix.go:18 +0x5c
created by github.com/docker/docker/daemon.(*Daemon).setupDumpStackTrap
	/go/src/github.com/docker/docker/daemon/debugtrap_unix.go:32 +0x100

goroutine 54 [chan receive]:
github.com/docker/docker/daemon/stats.(*Collector).Run(0xc420298c40)
	/go/src/github.com/docker/docker/daemon/stats/collector.go:62 +0x84
created by github.com/docker/docker/daemon.(*Daemon).newStatsCollector
	/go/src/github.com/docker/docker/daemon/stats_collector.go:24 +0xb4

goroutine 55 [chan receive]:
github.com/docker/docker/daemon.(*Daemon).execCommandGC(0xc42039a600)
	/go/src/github.com/docker/docker/daemon/exec.go:268 +0x70
created by github.com/docker/docker/daemon.NewDaemon
	/go/src/github.com/docker/docker/daemon/daemon.go:720 +0x2704

goroutine 56 [select]:
github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).watchLoop(0xc42000ae00)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/store.go:443 +0x114
created by github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).startWatch
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/store.go:460 +0x12c

goroutine 452 [runnable]:
github.com/docker/docker/vendor/github.com/hashicorp/memberlist.(*Memberlist).triggerFunc(0xc4214aa3c0, 0xbebc200, 0xc4214b2b40, 0xc4214b2a80, 0xc42126d400)
	/go/src/github.com/docker/docker/vendor/github.com/hashicorp/memberlist/state.go:119
created by github.com/docker/docker/vendor/github.com/hashicorp/memberlist.(*Memberlist).schedule
	/go/src/github.com/docker/docker/vendor/github.com/hashicorp/memberlist/state.go:106 +0x368

goroutine 429 [select]:
github.com/docker/docker/daemon.(*Daemon).DaemonLeavesCluster(0xc42039a600)
	/go/src/github.com/docker/docker/daemon/daemon.go:460 +0x190
github.com/docker/docker/daemon.(*Daemon).Shutdown(0xc42039a600, 0x0, 0x0)
	/go/src/github.com/docker/docker/daemon/daemon.go:858 +0x3b8
main.shutdownDaemon.func1(0xc42039a600, 0xc4212cfda0)
	/go/src/github.com/docker/docker/cmd/dockerd/daemon.go:390 +0x2c
created by main.shutdownDaemon
	/go/src/github.com/docker/docker/cmd/dockerd/daemon.go:392 +0x7c

goroutine 303 [select]:
github.com/docker/docker/vendor/github.com/docker/go-events.(*Broadcaster).run(0xc4212d6820)
	/go/src/github.com/docker/docker/vendor/github.com/docker/go-events/broadcast.go:117 +0x840
created by github.com/docker/docker/vendor/github.com/docker/go-events.NewBroadcaster
	/go/src/github.com/docker/docker/vendor/github.com/docker/go-events/broadcast.go:39 +0x198

goroutine 450 [runnable]:
github.com/docker/docker/vendor/github.com/hashicorp/memberlist.(*Memberlist).triggerFunc(0xc4214aa3c0, 0x3b9aca00, 0xc4214b2ae0, 0xc4214b2a80, 0xc42126d3f0)
	/go/src/github.com/docker/docker/vendor/github.com/hashicorp/memberlist/state.go:119
created by github.com/docker/docker/vendor/github.com/hashicorp/memberlist.(*Memberlist).schedule
	/go/src/github.com/docker/docker/vendor/github.com/hashicorp/memberlist/state.go:94 +0x1bc

goroutine 349 [IO wait]:
net.runtime_pollWait(0x3fffa004ce68, 0x72, 0xc420016160)
	/usr/local/go/src/runtime/netpoll.go:160 +0x8c
net.(*pollDesc).wait(0xc420a98a00, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x40
net.(*pollDesc).waitRead(0xc420a98a00, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x3c
net.(*netFD).accept(0xc420a989a0, 0x0, 0x12071560, 0xc4214b9880)
	/usr/local/go/src/net/fd_unix.go:419 +0x278
net.(*TCPListener).accept(0xc42042c870, 0xc4202ddef0, 0x0, 0x0)
	/usr/local/go/src/net/tcpsock_posix.go:132 +0x38
net.(*TCPListener).AcceptTCP(0xc42042c870, 0x1183ab20, 0x0, 0x0)
	/usr/local/go/src/net/tcpsock.go:209 +0xbc
github.com/docker/docker/vendor/github.com/hashicorp/memberlist.(*NetTransport).tcpListen(0xc420a98930, 0xc42042c870)
	/go/src/github.com/docker/docker/vendor/github.com/hashicorp/memberlist/net_transport.go:225 +0xa0
created by github.com/docker/docker/vendor/github.com/hashicorp/memberlist.NewNetTransport
	/go/src/github.com/docker/docker/vendor/github.com/hashicorp/memberlist/net_transport.go:109 +0x9d8

goroutine 351 [select]:
github.com/docker/docker/vendor/github.com/hashicorp/memberlist.(*Memberlist).streamListen(0xc4214aa3c0)
	/go/src/github.com/docker/docker/vendor/github.com/hashicorp/memberlist/net.go:190 +0x148
created by github.com/docker/docker/vendor/github.com/hashicorp/memberlist.newMemberlist
	/go/src/github.com/docker/docker/vendor/github.com/hashicorp/memberlist/memberlist.go:145 +0xc24

goroutine 152 [chan receive]:
main.(*DaemonCli).setupConfigReloadTrap.func1(0xc420af7b00, 0xc420489e30)
	/go/src/github.com/docker/docker/cmd/dockerd/daemon_unix.go:55 +0x50
created by main.(*DaemonCli).setupConfigReloadTrap
	/go/src/github.com/docker/docker/cmd/dockerd/daemon_unix.go:58 +0xf0

goroutine 350 [IO wait]:
net.runtime_pollWait(0x3fffa004cc28, 0x72, 0x0)
	/usr/local/go/src/runtime/netpoll.go:160 +0x8c
net.(*pollDesc).wait(0xc420a98ae0, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x40
net.(*pollDesc).waitRead(0xc420a98ae0, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x3c
net.(*netFD).readFrom(0xc420a98a80, 0xc421528000, 0x10000, 0x10000, 0x0, 0x0, 0x0, 0x120770a0, 0xc420016160)
	/usr/local/go/src/net/fd_unix.go:270 +0x244
net.(*UDPConn).readFrom(0xc42042c880, 0xc421528000, 0x10000, 0x10000, 0xc4202e1df0, 0x10000, 0x0, 0x0)
	/usr/local/go/src/net/udpsock_posix.go:43 +0x58
net.(*UDPConn).ReadFrom(0xc42042c880, 0xc421528000, 0x10000, 0x10000, 0x10000, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/udpsock.go:97 +0xf0
github.com/docker/docker/vendor/github.com/hashicorp/memberlist.(*NetTransport).udpListen(0xc420a98930, 0xc42042c880)
	/go/src/github.com/docker/docker/vendor/github.com/hashicorp/memberlist/net_transport.go:247 +0xf0
created by github.com/docker/docker/vendor/github.com/hashicorp/memberlist.NewNetTransport
	/go/src/github.com/docker/docker/vendor/github.com/hashicorp/memberlist/net_transport.go:110 +0xa54

goroutine 13 [IO wait]:
net.runtime_pollWait(0x3fffa004cfe8, 0x72, 0xc420016160)
	/usr/local/go/src/runtime/netpoll.go:160 +0x8c
net.(*pollDesc).wait(0xc4203f2300, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x40
net.(*pollDesc).waitRead(0xc4203f2300, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x3c
net.(*netFD).accept(0xc4203f22a0, 0x0, 0x12071560, 0xc42035a8c0)
	/usr/local/go/src/net/fd_unix.go:419 +0x278
net.(*UnixListener).accept(0xc420422400, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/unixsock_posix.go:158 +0x38
net.(*UnixListener).Accept(0xc420422400, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/unixsock.go:229 +0xc8
github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).acceptClientConnections(0xc42000ae00, 0xc42048c7e0, 0x5c, 0x120867a0, 0xc420422400)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/sandbox_externalkey_unix.go:128 +0x44
created by github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).startExternalKeyListener
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/sandbox_externalkey_unix.go:122 +0x27c

goroutine 456 [runnable]:
github.com/docker/docker/vendor/github.com/docker/libnetwork/networkdb.(*NetworkDB).triggerFunc(0xc4201431e0, 0x3b9aca00, 0xc4214b2d20, 0xc4214b2ba0, 0xc42126d450)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/networkdb/cluster.go:235
created by github.com/docker/docker/vendor/github.com/docker/libnetwork/networkdb.(*NetworkDB).clusterInit
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/networkdb/cluster.go:168 +0xbb0

goroutine 451 [runnable]:
github.com/docker/docker/vendor/github.com/hashicorp/memberlist.(*Memberlist).pushPullTrigger(0xc4214aa3c0, 0xc4214b2a80)
	/go/src/github.com/docker/docker/vendor/github.com/hashicorp/memberlist/state.go:141
created by github.com/docker/docker/vendor/github.com/hashicorp/memberlist.(*Memberlist).schedule
	/go/src/github.com/docker/docker/vendor/github.com/hashicorp/memberlist/state.go:100 +0x2a8

goroutine 352 [select]:
github.com/docker/docker/vendor/github.com/hashicorp/memberlist.(*Memberlist).packetListen(0xc4214aa3c0)
	/go/src/github.com/docker/docker/vendor/github.com/hashicorp/memberlist/net.go:270 +0x194
created by github.com/docker/docker/vendor/github.com/hashicorp/memberlist.newMemberlist
	/go/src/github.com/docker/docker/vendor/github.com/hashicorp/memberlist/memberlist.go:146 +0xc44

goroutine 194 [select]:
github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).clusterAgentInit(0xc42000ae00)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/controller.go:316 +0x33c
created by github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).SetClusterProvider
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/controller.go:252 +0xd4

goroutine 353 [select]:
github.com/docker/docker/vendor/github.com/hashicorp/memberlist.(*Memberlist).packetHandler(0xc4214aa3c0)
	/go/src/github.com/docker/docker/vendor/github.com/hashicorp/memberlist/net.go:352 +0x360
created by github.com/docker/docker/vendor/github.com/hashicorp/memberlist.newMemberlist
	/go/src/github.com/docker/docker/vendor/github.com/hashicorp/memberlist/memberlist.go:147 +0xc64

goroutine 348 [select]:
github.com/docker/docker/vendor/github.com/docker/go-events.(*Broadcaster).run(0xc4212d4e10)
	/go/src/github.com/docker/docker/vendor/github.com/docker/go-events/broadcast.go:117 +0x840
created by github.com/docker/docker/vendor/github.com/docker/go-events.NewBroadcaster
	/go/src/github.com/docker/docker/vendor/github.com/docker/go-events/broadcast.go:39 +0x198

goroutine 453 [runnable]:
github.com/docker/docker/vendor/github.com/docker/libnetwork/networkdb.(*NetworkDB).triggerFunc(0xc4201431e0, 0x12a05f200, 0xc4214b2c00, 0xc4214b2ba0, 0xc42126d420)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/networkdb/cluster.go:235
created by github.com/docker/docker/vendor/github.com/docker/libnetwork/networkdb.(*NetworkDB).clusterInit
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/networkdb/cluster.go:168 +0xbb0

goroutine 454 [runnable]:
github.com/docker/docker/vendor/github.com/docker/libnetwork/networkdb.(*NetworkDB).triggerFunc(0xc4201431e0, 0xbebc200, 0xc4214b2c60, 0xc4214b2ba0, 0xc42126d430)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/networkdb/cluster.go:235
created by github.com/docker/docker/vendor/github.com/docker/libnetwork/networkdb.(*NetworkDB).clusterInit
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/networkdb/cluster.go:168 +0xbb0

goroutine 402 [chan receive]:
main.(*DaemonCli).start.func2()
	/go/src/github.com/docker/docker/cmd/dockerd/daemon.go:258 +0x58
github.com/docker/docker/pkg/signal.Trap.func1.1(0xc4202fa17c, 0xc420304300, 0x1207cc60, 0xc4214157e0)
	/go/src/github.com/docker/docker/pkg/signal/trap.go:50 +0x190
created by github.com/docker/docker/pkg/signal.Trap.func1
	/go/src/github.com/docker/docker/pkg/signal/trap.go:65 +0x134

goroutine 301 [chan receive]:
github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).AgentInitWait(0xc42000ae00)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/controller.go:366 +0xac
github.com/docker/docker/daemon.(*Daemon).setupIngress(0xc42039a600, 0xc420d8d1a0, 0xc421290ad0, 0x10, 0x10, 0x0, 0x0)
	/go/src/github.com/docker/docker/daemon/network.go:163 +0x60
github.com/docker/docker/daemon.(*Daemon).startIngressWorker.func1(0xc42039a600)
	/go/src/github.com/docker/docker/daemon/network.go:122 +0xd0
created by github.com/docker/docker/daemon.(*Daemon).startIngressWorker
	/go/src/github.com/docker/docker/daemon/network.go:131 +0x74

goroutine 458 [semacquire]:
sync.runtime_notifyListWait(0xc4214a5410, 0xc400000000)
	/usr/local/go/src/runtime/sema.go:267 +0x200
sync.(*Cond).Wait(0xc4214a5400)
	/usr/local/go/src/sync/cond.go:57 +0xe8
github.com/docker/docker/vendor/github.com/docker/go-events.(*Queue).next(0xc421538990, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/github.com/docker/go-events/queue.go:103 +0x12c
github.com/docker/docker/vendor/github.com/docker/go-events.(*Queue).run(0xc421538990)
	/go/src/github.com/docker/docker/vendor/github.com/docker/go-events/queue.go:68 +0x38
created by github.com/docker/docker/vendor/github.com/docker/go-events.NewQueue
	/go/src/github.com/docker/docker/vendor/github.com/docker/go-events/queue.go:29 +0x1e8

goroutine 459 [runnable]:
github.com/docker/docker/vendor/github.com/docker/go-events.(*Queue).run(0xc421538a50)
	/go/src/github.com/docker/docker/vendor/github.com/docker/go-events/queue.go:66
created by github.com/docker/docker/vendor/github.com/docker/go-events.NewQueue
	/go/src/github.com/docker/docker/vendor/github.com/docker/go-events/queue.go:29 +0x1e8

goroutine 455 [runnable]:
github.com/docker/docker/vendor/github.com/docker/libnetwork/networkdb.(*NetworkDB).triggerFunc(0xc4201431e0, 0x6fc23ac00, 0xc4214b2cc0, 0xc4214b2ba0, 0xc42126d440)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/networkdb/cluster.go:235
created by github.com/docker/docker/vendor/github.com/docker/libnetwork/networkdb.(*NetworkDB).clusterInit
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/networkdb/cluster.go:168 +0xbb0

goroutine 457 [runnable]:
github.com/docker/docker/vendor/github.com/docker/libnetwork/networkdb.(*NetworkDB).triggerFunc(0xc4201431e0, 0x68c61714000, 0xc4214b2d80, 0xc4214b2ba0, 0xc42126d460)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/networkdb/cluster.go:235
created by github.com/docker/docker/vendor/github.com/docker/libnetwork/networkdb.(*NetworkDB).clusterInit
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/networkdb/cluster.go:168 +0xbb0

goroutine 460 [runnable]:
github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).handleTableEvents(0xc42000ae00, 0xc4214b2de0, 0xc42126d4b0)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/agent.go:734
created by github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).agentInit
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/agent.go:329 +0x604

goroutine 461 [runnable]:
github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).handleTableEvents(0xc42000ae00, 0xc4214b2f00, 0xc42126d4c0)
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/agent.go:734
created by github.com/docker/docker/vendor/github.com/docker/libnetwork.(*controller).agentInit
	/go/src/github.com/docker/docker/vendor/github.com/docker/libnetwork/agent.go:330 +0x670
```

--- a/cmd/dockerd/daemon.go
+++ b/cmd/dockerd/daemon.go
@@ -300,6 +300,11 @@ func (cli *DaemonCli) start(opts daemonOptions) (err error) {
 	if err != nil {
 		logrus.Fatalf("Error creating cluster component: %v", err)
 	}
+	d.SetCluster(c)
+	err = c.Start()
+	if err != nil {
+		logrus.Fatalf("Error starting cluster component: %v", err)
+	}
 
 	// Restart all autostart containers which has a swarm endpoint
 	// and is not yet running now that we have successfully
@@ -316,7 +321,6 @@ func (cli *DaemonCli) start(opts daemonOptions) (err error) {
 
 	cli.d = d
 
-	d.SetCluster(c)
 	initRouter(api, d, c)
 
 	cli.setupConfigReloadTrap()

--- a/daemon/cluster.go
+++ b/daemon/cluster.go
@@ -2,12 +2,14 @@ package daemon
 
 import (
 	apitypes "github.com/docker/docker/api/types"
+	lncluster "github.com/docker/libnetwork/cluster"
 )
 
 // Cluster is the interface for github.com/docker/docker/daemon/cluster.(*Cluster).
 type Cluster interface {
 	ClusterStatus
 	NetworkManager
+	SendClusterEvent(event lncluster.ConfigEventType)
 }
 
 // ClusterStatus interface provides information about the Swarm status of the Cluster

--- a/daemon/cluster/cluster.go
+++ b/daemon/cluster/cluster.go
@@ -51,6 +51,7 @@ import (
 	types "github.com/docker/docker/api/types/swarm"
 	executorpkg "github.com/docker/docker/daemon/cluster/executor"
 	"github.com/docker/docker/pkg/signal"
+	lncluster "github.com/docker/libnetwork/cluster"
 	swarmapi "github.com/docker/swarmkit/api"
 	swarmnode "github.com/docker/swarmkit/node"
 	"github.com/pkg/errors"
@@ -115,7 +116,7 @@ type Cluster struct {
 	root         string
 	runtimeRoot  string
 	config       Config
-	configEvent  chan struct{} // todo: make this array and goroutine safe
+	configEvent  chan lncluster.ConfigEventType // todo: make this array and goroutine safe
 	attachers    map[string]*attacher
 }
 
@@ -147,22 +148,30 @@ func New(config Config) (*Cluster, error) {
 	c := &Cluster{
 		root:        root,
 		config:      config,
-		configEvent: make(chan struct{}, 10),
+		configEvent: make(chan lncluster.ConfigEventType, 10),
 		runtimeRoot: config.RuntimeRoot,
 		attachers:   make(map[string]*attacher),
 	}
+	return c, nil
+}
+
+// Start the Cluster instance
+// TODO The split between New and Start can be join again when the SendClusterEvent
+// method is no longer required
+func (c *Cluster) Start() error {
+	root := filepath.Join(c.config.Root, swarmDirName)
 
 	nodeConfig, err := loadPersistentState(root)
 	if err != nil {
 		if os.IsNotExist(err) {
-			return c, nil
+			return nil
 		}
-		return nil, err
+		return err
 	}
 
 	nr, err := c.newNodeRunner(*nodeConfig)
 	if err != nil {
-		return nil, err
+		return err
 	}
 	c.nr = nr
 
@@ -172,10 +181,10 @@ func New(config Config) (*Cluster, error) {
 	case err := <-nr.Ready():
 		if err != nil {
 			logrus.WithError(err).Error("swarm component could not be started")
-			return c, nil
+			return nil
 		}
 	}
-	return c, nil
+	return nil
 }
 
 func (c *Cluster) newNodeRunner(conf nodeStartConfig) (*nodeRunner, error) {
@@ -308,7 +317,7 @@ func (c *Cluster) getRemoteAddressList() []string {
 // ListenClusterEvents returns a channel that receives messages on cluster
 // participation changes.
 // todo: make cancelable and accessible to multiple callers
-func (c *Cluster) ListenClusterEvents() <-chan struct{} {
+func (c *Cluster) ListenClusterEvents() <-chan lncluster.ConfigEventType {
 	return c.configEvent
 }
 
@@ -412,4 +421,14 @@ func (c *Cluster) lockedManagerAction(fn func(ctx context.Context, state nodeSta
 	defer cancel()
 
 	return fn(ctx, state)
+}
+
+// SendClusterEvent allows to send cluster events on the configEvent channel
+// TODO This method should not be exposed.
+// Currently it is used to notify the network controller that the keys are
+// available
+func (c *Cluster) SendClusterEvent(event lncluster.ConfigEventType) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	c.configEvent <- event
 }

--- a/daemon/cluster/noderunner.go
+++ b/daemon/cluster/noderunner.go
@@ -11,6 +11,7 @@ import (
 	"github.com/Sirupsen/logrus"
 	types "github.com/docker/docker/api/types/swarm"
 	"github.com/docker/docker/daemon/cluster/executor/container"
+	lncluster "github.com/docker/libnetwork/cluster"
 	swarmapi "github.com/docker/swarmkit/api"
 	swarmnode "github.com/docker/swarmkit/node"
 	"github.com/pkg/errors"
@@ -162,7 +163,7 @@ func (n *nodeRunner) handleControlSocketChange(ctx context.Context, node *swarmn
 		}
 		n.grpcConn = conn
 		n.mu.Unlock()
-		n.cluster.configEvent <- struct{}{}
+		n.cluster.SendClusterEvent(lncluster.EventSocketChange)
 	}
 }
 
@@ -175,7 +176,7 @@ func (n *nodeRunner) handleReadyEvent(ctx context.Context, node *swarmnode.Node,
 		close(ready)
 	case <-ctx.Done():
 	}
-	n.cluster.configEvent <- struct{}{}
+	n.cluster.SendClusterEvent(lncluster.EventNodeReady)
 }
 
 func (n *nodeRunner) handleNodeExit(node *swarmnode.Node) {
@@ -217,6 +218,7 @@ func (n *nodeRunner) Stop() error {
 	if err := n.swarmNode.Stop(ctx); err != nil && !strings.Contains(err.Error(), "context canceled") {
 		return err
 	}
+	n.cluster.SendClusterEvent(lncluster.EventNodeLeave)
 	<-n.done
 	return nil
 }

--- a/daemon/cluster/swarm.go
+++ b/daemon/cluster/swarm.go
@@ -388,7 +388,6 @@ func (c *Cluster) Leave(force bool) error {
 		}
 	}
 
-	c.configEvent <- struct{}{}
 	// todo: cleanup optional?
 	if err := clearPersistentState(c.root); err != nil {
 		return err

--- a/daemon/network.go
+++ b/daemon/network.go
@@ -16,6 +16,7 @@ import (
 	"github.com/docker/docker/pkg/plugingetter"
 	"github.com/docker/docker/runconfig"
 	"github.com/docker/libnetwork"
+	lncluster "github.com/docker/libnetwork/cluster"
 	"github.com/docker/libnetwork/driverapi"
 	"github.com/docker/libnetwork/ipamapi"
 	networktypes "github.com/docker/libnetwork/types"
@@ -207,7 +208,6 @@ func (daemon *Daemon) setupIngress(create *clustertypes.NetworkCreateRequest, ip
 
 func (daemon *Daemon) releaseIngress(id string) {
 	controller := daemon.netController
-
 	if err := controller.SandboxDestroy("ingress-sbox"); err != nil {
 		logrus.Errorf("Failed to delete ingress sandbox: %v", err)
 	}
@@ -233,13 +233,17 @@ func (daemon *Daemon) releaseIngress(id string) {
 		logrus.Errorf("Failed to delete ingress network %s: %v", n.ID(), err)
 		return
 	}
-
 	return
 }
 
 // SetNetworkBootstrapKeys sets the bootstrap keys.
 func (daemon *Daemon) SetNetworkBootstrapKeys(keys []*networktypes.EncryptionKey) error {
-	return daemon.netController.SetKeys(keys)
+	err := daemon.netController.SetKeys(keys)
+	if err == nil {
+		// Upon successful key setting dispatch the keys available event
+		daemon.cluster.SendClusterEvent(lncluster.EventNetworkKeysAvailable)
+	}
+	return err
 }
 
 // UpdateAttachment notifies the attacher about the attachment config.

--- a/integration-cli/docker_cli_swarm_test.go
+++ b/integration-cli/docker_cli_swarm_test.go
@@ -1980,3 +1980,24 @@ func (s *DockerSwarmSuite) TestSwarmInitUnspecifiedDataPathAddr(c *check.C) {
 	c.Assert(err, checker.NotNil)
 	c.Assert(out, checker.Contains, "data path address must be a non-zero IP")
 }
+
+func (s *DockerSwarmSuite) TestSwarmJoinLeave(c *check.C) {
+	d := s.AddDaemon(c, true, true)
+
+	out, err := d.Cmd("swarm", "join-token", "-q", "worker")
+	c.Assert(err, checker.IsNil)
+	c.Assert(strings.TrimSpace(out), checker.Not(checker.Equals), "")
+
+	token := strings.TrimSpace(out)
+
+	// Verify that back to back join/leave does not cause panics
+	d1 := s.AddDaemon(c, false, false)
+	for i := 0; i < 10; i++ {
+		out, err = d1.Cmd("swarm", "join", "--token", token, d.ListenAddr)
+		c.Assert(err, checker.IsNil)
+		c.Assert(strings.TrimSpace(out), checker.Not(checker.Equals), "")
+
+		_, err = d1.Cmd("swarm", "leave")
+		c.Assert(err, checker.IsNil)
+	}
+}

--- a/vendor.conf
+++ b/vendor.conf
@@ -25,7 +25,7 @@ github.com/imdario/mergo 0.2.1
 golang.org/x/sync de49d9dcd27d4f764488181bea099dfe6179bcf0
 
 #get libnetwork packages
-github.com/docker/libnetwork b015d4b1bcf4e666d8950651c8cc825a02842e7a
+github.com/docker/libnetwork 6786135bf7de08ec26a72a6f7e4291d27d113a3f
 github.com/docker/go-events 18b43f1bc85d9cdd42c05a6cd2d444c7a200a894
 github.com/armon/go-radix e39d623f12e8e41c7b5529e9a9dd67a1e2261f80
 github.com/armon/go-metrics eb0af217e5e9747e41dd5303755356b62d28e3ec

